### PR TITLE
refactor: use class in equals and hashCode methods instead of getCode

### LINKED ISSUE #784: Review notices class names and code
## **Feature request**

> We should review all rule class names and codes to make sure they follow the same naming pattern (this could be done in another PR). They should be the same, but there are a few that are different like this example - SameNameAndDescriptionForRouteNotice and same_route_name_and_description.

https://github.com/MobilityData/gtfs-validator/pull/781/files/67e5c7ad22e64c8bf00041baa2f28f21ba1c245b#discussion_r585866237

`UnexpectedEnumError` should be refactored to `UnexpectedEnumNotice`

> This notice should be refactored to UnexpectedEnumValueNotice because it's not an error, it's a warning.

https://github.com/MobilityData/gtfs-validator/pull/781#discussion_r585900910

### PR DESCRIPTION
**Pull request by lionel-nj**
closes #784

**Summary:**

This PR revisits the logic used to determine if two notices are equal, as well as their hashCode generation process.
- `equals`: uses the class simple instead of `getCode`
- `hashCode`: uses the class simple name instead of `getCode`

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- ~[ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)~

### CODE REVIEW

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.mobilitydata.gtfsvalidator.notice;
```
Comment (barbeau) [inline core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java:1]: Needs copyright header

**Review by barbeau — COMMENTED** [review]
Thanks @lionel-nj! Mostly LGTM, one small comment.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 2]
```diff
@@ -54,7 +54,7 @@ public boolean equals(Object other) {
     if (other instanceof Notice) {
       Notice otherNotice = (Notice) other;
       return context.equals(otherNotice.context)
-          && getCode().equals(otherNotice.getCode())
+          && getClass().getSimpleName().equals(otherNotice.getClass().getSimpleName())
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:57]: getClass().equals(otherNotice.getClass())
&& severityLevel.equals(otherNotice.severityLevel)
&& context.equals(otherNotice.context)

That is optimized for speed.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java [hunk 3]
```diff
@@ -71,6 +71,6 @@ public String toString() {
 
   @Override
   public int hashCode() {
-    return Objects.hash(getCode(), getContext(), getSeverityLevel());
+    return Objects.hash(getClass().getSimpleName(), getContext(), getSeverityLevel());
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java:74]: Objects.hash(getClass(), getContext(), getSeverityLevel());

I do not want to support use cases when you have classes with the same simple name but in different packages.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.notice;
+
+import java.util.Map;
+
+public class OtherTestValidationNotice extends ValidationNotice {
```
Comment (aababilov) [inline core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java:21]: Is that class reused in other tests? If not, please make it a static private class in the test where it is used.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  private final String code;
+
+  public OtherTestValidationNotice(
+      String code, Map<String, Object> context, SeverityLevel severityLevel) {
+    super(context, severityLevel);
+    this.code = code;
+  }
+
+  @Override
+  public String getCode() {
```
Comment (aababilov) [inline core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java:32]: I am unsure if you need that method. We plan to remove its implementations in child classes anyway.

**Review by aababilov — CHANGES_REQUESTED** [review]
I am on vacation will the end of this week. Please do you mind to hold on submitting changes to the upstream before I am back?

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java (hunk 4 above)
Comment (lionel-nj) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java:21]: It is not, will do. Thanks!

File: core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java (hunk 5 above)
Comment (lionel-nj) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/notice/OtherTestValidationNotice.java:32]: `getCode()` implementation is removed in https://github.com/MobilityData/gtfs-validator/pull/790 - but this PR has not been merged yet. I agree that in the future it should be removed.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 2 above)
Comment (lionel-nj) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java]: Modified in https://github.com/MobilityData/gtfs-validator/pull/789/commits/c0732911e4b070236fb67dd0ded790698de764e8. Thanks :)

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 3 above)
Comment (lionel-nj) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java]: Modified in https://github.com/MobilityData/gtfs-validator/pull/789/commits/c0732911e4b070236fb67dd0ded790698de764e8.

**Comment by lionel-nj** [comment]
Thanks for your review @aababilov! We will hold on submitting changes relates to the refactor of Notice's  `getCode()` method: #788, #789, #790, and #792.

**Comment by aababilov** [comment]
Good job, Lionel! Thank you!

I would be glad to submit this PR as-is. You can remove the `getCode()` function later.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java (hunk 2 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/Notice.java]: It still was not modified in a way that I suggested but you can submit it as-is.

**Review by aababilov — APPROVED** [review]

**Comment by lionel-nj** [comment]
Thanks for your review @aababilov and @barbeau  :)

**Pull request merged**